test/mb_debug_optimus.py

```diff
@@ -0,0 +1,667 @@
+    else:
+        observed_cell_molecule_gene_set.add((cell_barcode, molecule_barcode, gene_name))
+    
+    # find the indices that this molecule should correspond to
+    gene_index = gene_name_to_index[gene_name]
+
+    # if we've seen this cell before, get its index, else set it
+    try:
+        cell_index = cell_barcode_to_index[cell_barcode]
+    except KeyError:
+        cell_index = n_cells
+        cell_barcode_to_index[cell_barcode] = n_cells
+        n_cells += 1
+        
+    # record the molecule data
+    data.append(1)  # one count of this molecule
+    cell_indices.append(cell_index)
+    gene_indices.append(gene_index)
+    
+# convert into coo_matrix
+coordinate_matrix = sp.coo_matrix((data, (cell_indices, gene_indices)),
+    shape=(n_cells, n_genes), dtype=np.uint32)
+
+# convert into csr matrix and return
+col_index = np.asarray([k for k, v in sorted(gene_name_to_index.items(), key=operator.itemgetter(1))])
+row_index = np.asarray([k for k, v in sorted(cell_barcode_to_index.items(), key=operator.itemgetter(1))])
+# EPY: END code
+
+# EPY: START code
+print(f'optimus total counts: {np.sum(coordinate_matrix)}')
+# EPY: END code
+
+# EPY: START code
+import sctools
+import collections
+import tables
+import scipy.sparse as sp_sparse
+
+GeneBCMatrix = collections.namedtuple('GeneBCMatrix', ['gene_ids', 'gene_names', 'barcodes', 'matrix'])
+ 
+def get_matrix_from_h5(filename, genome):
+    with tables.open_file(filename, 'r') as f:
+        try:
+            group = f.get_node(f.root, genome)
+        except tables.NoSuchNodeError:
+            print("That genome does not exist in this file.")
+            return None
+        gene_ids = getattr(group, 'genes').read()
+        gene_names = getattr(group, 'gene_names').read()
+        barcodes = getattr(group, 'barcodes').read()
+        data = getattr(group, 'data').read()
+        indices = getattr(group, 'indices').read()
+        indptr = getattr(group, 'indptr').read()
+        shape = getattr(group, 'shape').read()
+        matrix = sp_sparse.csc_matrix((data, indices, indptr), shape=shape)
+        return GeneBCMatrix(gene_ids, gene_names, barcodes, matrix)
+ 
+filtered_matrix_h5 = "cell_ranger/raw_gene_bc_matrices_h5.h5"
+genome = "GRCh38"
+cr_cm = get_matrix_from_h5(filtered_matrix_h5, genome)
+# EPY: END code
+
+# EPY: START code
+print(f'cell ranger total counts: {cr_cm.matrix.sum()}')
+# EPY: END code
+
+# EPY: START markdown
+# **Summary**
+# 
+# - Close enough -- 95% concordance.
+# - I do not quite understand the difference; some thoughts:
+#   - Duplicate marking is immaterial for 10x since we only count unique (CB, GE, UB) tuples.
+#   - Multiple primary alignment is not an issue (unles CellRanger throws away reads with multiple primary alignments?)
+#   - Filtering by MQ could be an issue (Optimus does not filter, CellRanger does, but it manipulates the STAR bam out such that it is difficult to interpret their downstream filter MQ == 255)
```

Comment:
Looks like I never merged the multi-mapping changes. There is an intention to filter. 

https://github.com/HumanCellAtlas/sctools/commit/c81eb0a1f4634ca4116be3461e88a7c29f4b9f3e

Do the number of reads in Cell Ranger with MAPQ < 255 explain the difference in molecule number? If so, this could explain why we see more molecules, since we're currently not filtering.